### ISSUE TITLE
NamedNodeMap.getNamedItem('style') is `null` in IE11

### ISSUE THREAD

**Issue by emilio-martinez** [Polymer/lit-html#425]
I'm seeing an error emitted on IE11 that I've traced to lit-html's `Template` where `attributes.getNamedItem(attributeNameInPart)` returns `null` when `attributeNameInPart` is `'style'`. I haven't experienced this with any other attributes (yet), which is why I'm relating this specifically to the `style` attribute until proven otherwise. 

The relevant template snippet I have is:

```html
<div class="swatch__color" style="background-color: ${hex};">
```

Perhaps worth noting is that I'm using `lit-html` via `lit-element`, but I was able to reproduce the same issue using `lit-html` on its own. I've only seen this in IE11 so far—I've tested on Safari, Chrome, and Edge 16. Everything works fine otherwise (with polyfills, of course) if I remove the style attribute in the example above.

<img width="1504" alt="lit-html-ie11-namednodemap-null" src="https://user-images.githubusercontent.com/9163179/43694790-f585c2dc-98e9-11e8-8d2a-b30cb8934613.png">

**Comment by emilio-martinez** [comment]
Looking at the attribute syntax in the docs I also tried the below, but got the same error

```html
<div class="swatch__color" style=${`background-color: ${hex};`}>
```

**Comment by csvn** [comment]
I've been testing out `lit-html@dev` a bit, which has the updated parts API. The issue occurs there too. I made a small reproduction to try it out:

https://stackblitz.com/edit/js-womx5n

![image](https://user-images.githubusercontent.com/8770194/43858890-8001fa40-9b4f-11e8-8827-ece95ae7b28a.png)

When I compared the HTML of the template before it was "prepared", it seems IE 11 removes the `{{lit-XXX}}` placeholder inside the `style` attribute:

```html
<!-- Chrome -->
<div class="box" style="{{lit-22822797467408895}}"><!--{{lit-22822797467408895}}--></div>

<!-- IE 11 -->
<div class="box" style=""><!--{{lit-41187566288397953}}--></div>
```

This is why the value for the `style` attribute ends up displayed as text in IE, since the binding is skipped.

**Comment by ruphin** [comment]
This is a really awkward part of IE11; it interprets and rewrites the `style` attribute value based on what is valid CSS. If you write something that is invalid CSS, it will be ignored. 

A solution is to change the shape of the attribute marker to something that parses as valid CSS in IE11. I found that IE11 accepts basically any value in `font-family`, which is convenient because it lets us keep a random component in the marker. Unfortunately, it also adds a space after the `:` character, which makes checking the value difficult because it is transformed.

When using this as a marker
```javascript
const attributeMarker = `font-family:<random-string>`;
```

IE11 will render this DOM
```
style="font-family: <random-string>"
``` 
and other browsers will render 
```
style="font-family:<random-string>"
```

When there is no way around this, it might be easier to consider `style` a special case and test for it specifically. With some preliminary experimentation I have not found a nice way of doing this yet that does not conflict with detecting other edge cases such as #403

I will update when I find a solution

**Comment by ruphin** [comment]
The solution I described above works, in a roundabout way. When using `font-family:<random-string>` as a marker for attributes, you can detect its presence in IE11:

```javascript
// Globals
const attributeMarker = `font-family:${marker}`;
const IEStyleMarker = `font-family: ${marker};`;

const attributes = node.attributes;
let count = 0;
for (let i = 0; i < attributes.length; i++) {
  if (attributes[i].value.indexOf(attributeMarker) >= 0) {
    count++;
  }
}
if (node.getAttribute('style').indexOf(IEStyleMarker) >= 0) {
  count++;
}
```

In my opinion it's cleaner with ES6. This does allocate an additional array for every invocation, but if you skip nodes that do not have dynamic attributes (which you can), that cost is very negligible, and it saves ~10 Bytes after minification (~70 Bytes when compiled to ES5):

```javascript
// Globals
const filter = [].filter;
const attributeMarker = `font-family:${marker}`;
const IEStyleMarker = `font-family: ${marker};`;


let count = filter.call(node.attributes, 
     attribute => attribute.value.indexOf(attributeMarker) >= 0
).length;
if (node.getAttribute('style').indexOf(IEStyleMarker) >= 0) {
  count += 1;
}
```
Note that `marker` here may only contain alphanumerics. 

The additional `if` will never cause the `style` attribute to be counted twice, because other browsers do not inject the space character.

**Comment by justinfagnani** [comment]
This should have been fixed by #448 

Please reopen if it's not.

**Comment by limonte** [comment]
> Please reopen if it's not.

FYI, a reported can't reopen the issue. https://stackoverflow.com/a/21333938
